mapie/calibrators/ccp/utils.py

```diff
@@ -0,0 +1,534 @@
+    """
+    if isinstance(sigma, (float, int)):
+        return np.ones((n_points, 1))*sigma
+    else:
+        if len(np.array(sigma).shape) != 1:
+            raise ValueError("sigma argument should be a float "
+                             "or a 1D array of floats.")
+        return np.ones((n_points, 1))*np.array(sigma)
+
+
+def dynamic_arguments_call(f: Callable, params_mapping: Dict) -> NDArray:
+    """
+    Call the function ``f``, with the correct arguments
+
+    Parameters
+    ----------
+    f : Callable
+        function to call
+
+    params_mapping : Dict
+        Dictionnary of argument names / values
+
+    Returns
+    -------
+    NDArray
+        result as 2D array
+    """
+
+    params = inspect.signature(f).parameters
+    used_params = {
+        p: params_mapping[p] for p in params
+        if p in params_mapping and params_mapping[p] is not None
+    }
+    res = np.array(f(**used_params), dtype=float)
+    if len(res.shape) == 1:
+        res = np.expand_dims(res, axis=1)
+
+    return res
+
+
+def concatenate_functions(
+    functions: List[Callable], params_mapping: Dict,
+    multipliers: Optional[List[Callable]]
+) -> NDArray:
+    """
+    Call the function of ``functions``, with the
+    correct arguments, and concatenate the results
+
+    Parameters
+    ----------
+    functions : List[Callable]
+        List of functions to call
+
+    params_mapping : Dict
+        Dictionnary of argument names / values
+
+    Returns
+    -------
+    NDArray
+        Concatenated result
+    """
+    # Compute phi(X, y_pred, z)
+    result = np.hstack([
+        dynamic_arguments_call(f, params_mapping) for f in functions
+    ])
+    # Multiply the result by each multiplier function
+    if multipliers is not None:
+        for f in multipliers:
+            result *= dynamic_arguments_call(f, params_mapping)
+    return result
+
+
+def check_multiplier(
+    multipliers: Optional[List[Callable]],
+    X: Optional[ArrayLike] = None,
+    y_pred: Optional[ArrayLike] = None,
+    z: Optional[ArrayLike] = None,
+) -> None:
+    """
+    Check if ``multipliers`` is a valid ``multiplier`` argument for
+    ``CCPCalibrator``.
+
+    Parameters
+    ----------
+    multipliers : List[Callable]
+        function which sould return an array of shape (n_samples, 1) or
+        (n_samples, )
+
+    X : ArrayLike
+            Observed samples
+
+    y_pred : ArrayLike
+        Target prediction
+
+    z : ArrayLike
+        Exogenous variable
+    """
+    if multipliers is None:
+        return
+    params_mapping = {"X": X, "y_pred": y_pred, "z": z}
+    for f in multipliers:
+        res = dynamic_arguments_call(f, params_mapping)
+        if res.shape != (_num_samples(X), 1):
+            raise ValueError("The function used as multiplier should return an"
+                             "array of shape n_samples, 1) or (n_samples, ).\n"
+                             f"Got shape = {res.shape}.")
+
+
+def fast_mean_pinball_loss(
```

Comment:
Could this be used in `MapieQuantileRegressor` to turn any regressor into a quantile regressor?

Reply:
I guess so. Assuming the estimators all have an argument to change the default loss.

Reply:
Cool, I think it would be worth maybe creating a new issue to report this.

Reply:
Aren't we going to redo all the ``CQR`` class in Maple V1 ? :)